### PR TITLE
Fix WFS 2.0.0 combined with OGC HTTP URI srsName style

### PR DESCRIPTION
**Pull request by dollarklavs**
## Description
Bug report #21575: Added new 'else if' branch that handles OGC HTTP URL srsName style, so that epsgNr can be extracted from srsName in the QgsGmlStreamingParser::readEpsgFromAttribute.

Also added a another new 'else if' branch for QgsCoordinateReferenceSystem::createFromOgcWmsCrs which match autority and code from the srsName
[#21575](https://issues.qgis.org/issues/21575)

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by elpaso** [comment]
Looks good, can you please add a test case?
The changes are in core classes, they need to be protected by test cases.

**Comment by dollarklavs** [comment]
Hi @elpaso,
Is there an existing test for URN style `srsName`'s that I can base this test on? I Don't see any existing tests for WFS under providers - are there any?

**Comment by elpaso** [comment]
@dollarklavs your changes are not in WFS provider but in core, look at testqgscoordinatereferencesystem.cpp and testqgsogcutils.cpp and testqgsgml.cpp for examples.

**Comment by elpaso** [comment]
@dollarklavs forgot to mention that we also have some WFS test in tests/src/python/

**Comment by dollarklavs** [comment]
@elpaso I have added tests in testqgscoordinatereferencesystem.cpp and testqgsgml.cpp.

I'm unsure if I should add more tests, as, if I follow the current convention, would have to duplicate a lot of code in  tests/src/python/test_provider_wfs.py.
What do you say?

File: src/core/qgscoordinatereferencesystem.cpp [hunk 1]
```diff
@@ -333,15 +333,20 @@ bool QgsCoordinateReferenceSystem::createFromOgcWmsCrs( const QString &crs )
 
   QString wmsCrs = crs;
 
-  QRegExp re( "urn:ogc:def:crs:([^:]+).+([^:]+)", Qt::CaseInsensitive );
-  if ( re.exactMatch( wmsCrs ) )
+  QRegExp reUri( "http://www\\.opengis\\.net/def/crs/([^/]+).+/(\\d+)", Qt::CaseInsensitive );
```
Comment (elpaso) [inline src/core/qgscoordinatereferencesystem.cpp]: You may want to use raw strings here for readability (avoid double slashes):

**Comment by elpaso** [comment]
@dollarklavs I think it's fine. 

@rouault can you also have a look please?

File: src/core/qgscoordinatereferencesystem.cpp (hunk 1 above)
Comment (dollarklavs) [reply src/core/qgscoordinatereferencesystem.cpp]: @elpaso -sorry for the late responce. Were you thinking of something like this?
`QRegExp reUri( R"(http://www\.opengis\.net/def/crs/([^/]+).+/(\d+))", Qt::CaseInsensitive );`

Comment (elpaso) [reply src/core/qgscoordinatereferencesystem.cpp]: Yes, that should work (untested)

**Comment by nyalldawson** [comment]
@dollarklavs can you please rebase this and fix the merge conflicts?

**Comment by dollarklavs** [comment]
@nyalldawson didn't see your comment until now - sorry.
Can you give me a pointer as to how I rebase. The commands that I have tried just says 

> Current branch OGC_HTTP_URI_fix is up to date.

**Comment by dollarklavs** [comment]
@nyalldawson I have fixed the merge conflicts and the fix passes all tests it seems

**Comment by DelazJ** [comment]
@nyalldawson Just a reminder: It seems (discussion on gitter) that this fix was merged in master but not backported to 3.8 (despite it was tagged for 3.8 branch). Any issue or an omission?

**Comment by nyalldawson** [comment]
@DelazJ it's not up to me to ensure other's bug fixes are backported, that responsibility sits with the PR submitter.

**Comment by DelazJ** [comment]
With the backport labels, the idea is to avoid anyone having to do it manually, unless there's conflict. And not all submitters have rights to add labels to their PR (or issue report)

**Comment by nyalldawson** [comment]
> With the backport labels, the idea is to avoid anyone having to do it manually, unless there's conflict. 

No, that's not true. It's there just to make backports simple where it's appropriate. The responsibility still lies clearly on the person making it a backport to ensure the fix is valid in the other branches and isn't causing regressions. Just adding the tag isn't safe - someone has to review the change's applicability to the stable branches and personally put their name to ensuring it's ok to do so.

If we just throw the label on all PRs and rely on Travis to pick up errors, it's going to cause breakage!

I'll add the tag myself where I think it's safe to do so, and I'm willing to put my name and reputation behind the backport. I won't always do this, and would rather play it safe and just not backport stuff instead of risking this. (Especially given recent hostility towards regressions in stable releases and the consequences of backporting stuff. It's just not worth the stress.)

This is an open project, and anyone can submit backports via PRs, if they've tested and verified that they are safe to do.

**Comment by DelazJ** [comment]
I don't know the implication of the changes in this PR so take my comment as a general one (regarding those tags management) based on my experience/expectation and not specific to this PR.
 > > With the backport labels, the idea is to avoid anyone having to do it manually, unless there's conflict.
> 
> No, that's not true. It's there just to make backports simple where it's appropriate. The responsibility still lies clearly on the person making it a backport to ensure the fix is valid in the other branches and isn't causing regressions. Just adding the tag isn't safe - someone has to review the change's applicability to the stable branches and personally put their name to ensuring it's ok to do so.
> 
> If we just throw the label on all PRs and rely on Travis to pick up errors, it's going to cause breakage!

Indeed, we need to ensure that every commit is safe before merging, and this check might belong to the submitter (though I think/hope reviewers also do that; submitting a PR with a green Travis is not enough to get it merged, is it?).

> I'll add the tag myself where I think it's safe to do so, and I'm willing to put my name and reputation behind the backport. I won't always do this, and would rather play it safe and just not backport stuff instead of risking this. 

Regarding the tag, afaics, adding it (should it be on submitter's request? somehow it's the case here) does not put the core dev name on the commits, they appear nowhere in the history, nor in the backport PR. The initial author is still the only one mentioned. If there are other responsible, they would be the reviewers/"merger" but I don't think that the situation would then be different from a manually submitted PR.
I however see where you seem to find your responsibility engaged with enabling backporting so maybe a formal request from the author would help here. In doc repo, we have a checkbox in PR template whether the fix needs to be backported or not. And depending on its state, we add the label or not. Maybe adding such checkbox for LTR and for LR (with clearer statement) could do the trick here and clearly "engage" PR author...

> (Especially given recent hostility towards regressions in stable releases and the consequences of backporting stuff. It's just not worth the stress.)

Are there true regressions? :roll_eyes: 
> This is an open project, and anyone can submit backports via PRs, if they've tested and verified that they are safe to do.

Correct! But since we have tools that ease the process, let's them benefit to all. And saying that I particularly have in mind PRs I submitted (with few impact) #30462, #31025 or #30308. Should I have (had) to (branch and) checkout the release branch(es), cherry-pick the commit, push and create a PR (for each) while adding the tag(s) would allow me to spend that time on hiking or submitting other fixes? Labels undoubtedly save time to the submitter.
My point is that if the fix looks safe (and core devs + travis are likely more aware of that) and is compatible with backport rules, let's label it (unless we request author agreement beforehand, hence implement the above idea).

**Pull request merged**